File: mobile-flutter/lib/data/services/auth_service.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../../core/config/api_config.dart';
import '../models/user_model.dart';
import '../models/auth_models.dart';

class AuthService {
  static const String _tokenKey = 'access_token';
  static const String _userKey = 'user_data';
  static const String _refreshTokenKey = 'refresh_token';
  
  final Dio _dio = Dio(BaseOptions(
    baseUrl: ApiConfig.userApiUrl,
    connectTimeout: ApiConfig.connectTimeout,
    receiveTimeout: ApiConfig.receiveTimeout,
    sendTimeout: ApiConfig.sendTimeout,
    headers: ApiConfig.defaultHeaders,
  ));

  final FlutterSecureStorage _secureStorage = const FlutterSecureStorage(
    aOptions: AndroidOptions(
      encryptedSharedPreferences: true,
    ),
    iOptions: IOSOptions(
      accessibility: KeychainItemAccessibility.first_unlock_this_device,
    ),
  );

  AuthService() {
    _setupInterceptors();
  }

  void _setupInterceptors() {
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) async {
          // Add token to requests automatically
          final token = await getToken();
          if (token != null) {
            options.headers['Authorization'] = 'Bearer $token';
          }
          handler.next(options);
        },
        onError: (error, handler) async {
          // Handle 401 errors with token refresh
          if (error.response?.statusCode == 401) {
            final refreshed = await _tryRefreshToken();
            if (refreshed) {
              // Retry the original request
              final clonedRequest = await _dio.fetch(error.requestOptions);
              handler.resolve(clonedRequest);
              return;
            }
          }
          handler.next(error);
        },
      ),
    );
  }

  Future<LoginResponse> login(LoginRequest request) async {
    try {
      final response = await _dio.post(
        ApiConfig.endpoints['login']!,
        data: request.toJson(),
      );

      final loginResponse = LoginResponse.fromJson(response.data);
      
      // Store tokens securely
      await _secureStorage.write(key: _tokenKey, value: loginResponse.accessToken);
      
      if (loginResponse.refreshToken != null) {
        await _secureStorage.write(key: _refreshTokenKey, value: loginResponse.refreshToken!);
      }

      // Create and store user data since API might not return it
      final user = User(
        id: DateTime.now().millisecondsSinceEpoch.toString(),
        username: request.username,
        email: '${request.username}@company.com',
        role: request.username.toLowerCase() == 'admin' ? 'Administrator' : 'User',
        isActive: true,
        createdAt: DateTime.now(),
        updatedAt: DateTime.now(),
        lastLogin: DateTime.now(),
      );

      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_userKey, jsonEncode(user.toJson()));

      return loginResponse;
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw AuthException('Invalid credentials');
      } else if (e.response?.statusCode == 422) {
        throw AuthException('Invalid input format');
      } else if (e.type == DioExceptionType.connectionTimeout) {
        throw AuthException('Connection timeout');
      } else if (e.type == DioExceptionType.receiveTimeout) {
        throw AuthException('Server response timeout');
      }
      throw AuthException('Login failed: ${e.message}');
    } catch (e) {
      throw AuthException('Unexpected error: ${e.toString()}');
    }
  }

  Future<void> logout() async {
    try {
      // Attempt to logout on server
      await _dio.post(ApiConfig.endpoints['logout']!);
    } catch (e) {
      // Continue with local logout even if server logout fails
    } finally {
      // Clear all stored data
      await _secureStorage.delete(key: _tokenKey);
      await _secureStorage.delete(key: _refreshTokenKey);
      
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove(_userKey);
    }
  }

  Future<String?> getToken() async {
    return await _secureStorage.read(key: _tokenKey);
  }

  Future<User?> getCurrentUser() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final userJson = prefs.getString(_userKey);
      
      if (userJson != null) {
        return User.fromJson(jsonDecode(userJson));
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  Future<bool> isAuthenticated() async {
    final token = await getToken();
    return token != null && token.isNotEmpty;
  }

  Future<bool> _tryRefreshToken() async {
    try {
      final refreshToken = await _secureStorage.read(key: _refreshTokenKey);
      if (refreshToken == null) return false;

      final response = await _dio.post(
        ApiConfig.endpoints['refresh']!,
        data: {'refresh_token': refreshToken},
      );

      final newToken = response.data['access_token'] as String?;
      if (newToken != null) {
        await _secureStorage.write(key: _tokenKey, value: newToken);
        return true;
      }
      return false;
    } catch (e) {
      return false;
    }
  }
}

class AuthException implements Exception {
  final String message;
  AuthException(this.message);

  @override
  String toString() => message;
}

File: mobile/src/services/authService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { API_CONFIG } from '../config/api';
import { LoginRequest, LoginResponse, User } from '../types/index';

class AuthService {
  private baseURL = API_CONFIG.USER_API_URL;

  async login(credentials: LoginRequest): Promise<LoginResponse> {
    try {
      console.log('Attempting login to:', `${this.baseURL}${API_CONFIG.ENDPOINTS.AUTH.LOGIN}`);
      
      const response = await fetch(`${this.baseURL}${API_CONFIG.ENDPOINTS.AUTH.LOGIN}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(credentials),
      });

      console.log('Login response status:', response.status);

      if (!response.ok) {
        const errorData = await response.json();
        console.error('Login failed with error:', errorData);
        throw new Error(errorData.detail || 'Login failed');
      }

      const data: LoginResponse = await response.json();
      console.log('Login successful, received data:', data);
      
      // Store token first
      await AsyncStorage.setItem('access_token', data.access_token);
      
      // Create user data based on login credentials since API doesn't return user info
      const userData: User = {
        id: 'user_' + Date.now(), // Generate a simple ID
        username: credentials.username,
        email: credentials.username.includes('@') ? credentials.username : `${credentials.username}@company.com`,
        role: credentials.username === 'admin' ? 'admin' : 'viewer',
        is_active: true,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString()
      };
      
      console.log('Storing user data:', userData);
      await AsyncStorage.setItem('user_data', JSON.stringify(userData));
      
      return {
        ...data,
        user: userData
      };
    } catch (error) {
      console.error('Login error:', error);
      // Provide more specific error messages
      if (error instanceof TypeError && error.message.includes('Network request failed')) {
        throw new Error('Cannot connect to server. Please check if the backend is running and your network connection.');
      }
      throw error;
    }
  }

  async logout(): Promise<void> {
    try {
      const token = await this.getToken();
      if (token) {
        await fetch(`${this.baseURL}${API_CONFIG.ENDPOINTS.AUTH.LOGOUT}`, {
          method: 'POST',
          headers: {
            'Authorization': `Bearer ${token}`,
          },
        });
      }
    } catch (error) {
      console.error('Logout error:', error);
    } finally {
      // Clear local storage regardless of API call result
      await AsyncStorage.removeItem('access_token');
      await AsyncStorage.removeItem('user_data');
    }
  }

  async getToken(): Promise<string | null> {
    return await AsyncStorage.getItem('access_token');
  }

  async getCurrentUser(): Promise<User | null> {
    try {
      const userData = await AsyncStorage.getItem('user_data');
      return userData ? JSON.parse(userData) : null;
    } catch (error) {
      console.error('Error getting current user:', error);
      return null;
    }
  }

  async isAuthenticated(): Promise<boolean> {
    const token = await this.getToken();
    return token !== null;
  }

  async refreshToken(): Promise<string | null> {
    try {
      const token = await this.getToken();
      if (!token) return null;

      const response = await fetch(`${this.baseURL}${API_CONFIG.ENDPOINTS.AUTH.REFRESH}`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${token}`,
        },
      });

      if (response.ok) {
        const data = await response.json();
        await AsyncStorage.setItem('access_token', data.access_token);
        return data.access_token;
      }
    } catch (error) {
      console.error('Token refresh error:', error);
    }
    
    return null;
  }
}

export default new AuthService();
